### PR TITLE
Add a new scene and connection code to receive HoloLens poses from the camera

### PR DESCRIPTION
**Pull request by matthejo**
This change has a few main parts:
1. A new scene (SpectatorViewCamera) that represents the application that runs on the holographic camera rig. In this PR, the app provides time-adjusted poses from the HoloLens to the compositor and transfers a JSON blob of calibration data. This data is not understood by the SpectatorViewCamera scene, it's up to the compositor to parse and understand that data blob.
2. UI has been added to the Compositor window to connect to the SpectatorViewCamera app using TCPConnectionManager.
3. There's a registry of calibration parsers (currently there are zero checked in) that can receive calibration data and attempt to parse it into an ICalibrationData.

With an internal calibration I've tested the end-to-end workflow of connecting the camera rig to the compositor and filming a cube floating in space using the rig.

### CODE REVIEW

**Comment by andreiborodin** [comment]
>         private ICalibrationData calibrationData;

Do you not get "unused variable" warning when building non-editor?

---
Refers to: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/CompositionManager.cs:80 in a352b1c. [](commit_id = a352b1ce0850b8941186f7deaf247c3db56552bb, deletion_comment = False)

File: Assets/MixedRealityToolkit.Extensions/Socketer/Scripts/NetworkExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.Socketer
+{
+    public static class NetworkExtensions
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/Socketer/Scripts/NetworkExtensions.cs]: >public [](start = 4, length = 6)

nit: please make this internal

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/CompositionManager.cs [hunk 2]
```diff
@@ -452,28 +462,27 @@ protected void OnPreCull()
...
             this.calibrationData = calibrationData;
-            GameObject container = GameObject.Find("SpectatorView");
-            if (container == null)
+            if (videoCameraPose == null)
             {
-                container = GameObject.CreatePrimitive(PrimitiveType.Cube);
-                container.GetComponent<MeshRenderer>().enabled = false;
-                GameObject.DestroyImmediate(container.GetComponent<Collider>());
-                container.name = "SpectatorView";
+                videoCameraPose = new GameObject("HoloLens Pose");
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/CompositionManager.cs]: >HoloLens [](start = 50, length = 8)

nit: may want to call this hmd compared to hololens?

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    /// </summary>
+    public interface ICalibrationParser
+    {
+        /// <summary>
+        /// Tries to parse the provided string as calibration data.
+        /// </summary>
+        /// <param name="rawCalibrationData">The string representation of calibration data to deserialize.</param>
+        /// <param name="calibrationData">The parsed calibration data, which can be used to configure a stereo camera rig.</param>
+        /// <returns>True if the string was successfully parsed, otherwise false.</returns>
+        bool TryParse(string rawCalibrationData, out ICalibrationData calibrationData);
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs]: >string [](start = 22, length = 6)

I'm not sure that we should assume that calibration data will be handed around in a string. Would it be possible to make this a byte array instead?

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/HolographicCameraNetworkManager.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                            Vector3 cameraPosition = reader.ReadVector3();
+                            Quaternion cameraRotation = reader.ReadQuaternion();
+
+                            compositionManager.AddCameraPose(cameraPosition, cameraRotation, timestamp);
+                        }
+                        break;
+                    case "CalibrationData":
+                        {
+                            string calibrationDataJson = reader.ReadString();
+                            CalibrationPackage calibrationPackage = JsonUtility.FromJson<CalibrationPackage>(calibrationDataJson);
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/HolographicCameraNetworkManager.cs]: >FromJson [](start = 80, length = 8)

nit: please add this serialization logic as helper functions in the class.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView.Compositor
+{
+    /// <summary>
+    /// Interface to allow multiple parsers to attempt to parse calibration data.
+    /// </summary>
+    public interface ICalibrationParser
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs]: >ICalibrationParser [](start = 21, length = 18)

does this end up as a monobehaviour? it seems like maybe we should add an example using the CalibrationData class that is now in this repo.

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/HolographicCamera/HolographicCameraNetworkListener.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView.HolographicCamera
+{
+    /// <summary>
+    /// Represents the network listener for the camera pose provider app.
+    /// </summary>
+    [RequireComponent(typeof(TCPConnectionManager))]
+    public class HolographicCameraNetworkListener : MonoBehaviour
```
Comment (chrisfromwork) [inline Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/HolographicCamera/HolographicCameraNetworkListener.cs]: >HolographicCameraNetworkListener [](start = 17, length = 32)

This isn't really camera specific, we may just want to call this network listener

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs (hunk 5 above)
Comment (matthejo) [reply Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs]: Will definitely get to this but not with this change. Need to get an example calibration from Chris to test ingesting it with CalibrationData

File: Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs (hunk 3 above)
Comment (matthejo) [reply Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Compositor/ICalibrationParser.cs]: For the calibration data to be JSON, it either needs to always be a string, or be stored in the JSON blob as a byte[].  Right now I've opted for a string but we can revisit this later.

**Pull request merged**